### PR TITLE
WIP: Profiles plugin

### PR DESCRIPTION
**Pull request by Thra11**
I've been thinking about how, depending on whether it's day or night, I adjust a number of settings, but I usually end up with one of two states:

Setting | Dark | Light
--------- | ------ | -----
Frontlight Brightness | 1 | 16
Frontlight Warmth | 100 | 30
Night Mode | On | Off
Font Contrast | Higher | Low

My idea to make this more convenient is as follows.
* The user is able to configure two (possibly more) "profiles".
* With a UI similar to the gesture configuration UI, the user can choose multiple settings (with values) to add to the profile.
* Loading a profile simply sets the chosen settings to the selected values (After setting the settings, KOReader doesn't need to remember which profile was loaded, and the user can adjust the settings as usual).
* An action to toggle (or cycle) between profiles could be made available, so a gesture can toggle between profiles.

Would probably also solve some other issues, such as #1427 & #5019.

So far, this is a proof-of-concept which allows you to use gestures to "cycle through profiles" (execute one of two lua files). At the moment, the "profiles" are hand-written scripts (copy them into koreader/profiles/ to use them): It doesn't have a GUI to select desired settings and values yet.

Remarks
I thought about trying to hook into the gesture manager's actions, as a way to have a selection of relevant actions available instead of arbitrary code execution. However, actions are too limited for my use case, as they don't allow you to select custom values (e.g. there's a "toggle frontlight" action, but there's no "set frontlight intensity to _x_", nor support in the GUI for choosing _x_. Similarly, there's a "toggle night mode", but no "turn night mode _on_", etc.).

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/5079)

### CODE REVIEW

File: plugins/profiles.koplugin/profile_1.lua [hunk 1]
```diff
@@ -0,0 +1,14 @@
+local Device = require("device")
+local Screen = require("device").screen
```
Comment (Frenzie) [inline plugins/profiles.koplugin/profile_1.lua]: It doesn't make much practical difference but according to the coding style that should be:
```suggestion
local Screen = Device.screen
```

File: plugins/profiles.koplugin/main.lua [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+local number_of_profiles = 2
+
+function Profiles:init()
+    logger.info("Profiles:init()")
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function Profiles:addToMainMenu(menu_items)
+    logger.info("Profiles:addToMainMenu")
+    menu_items.profiles = {
```
Comment (Frenzie) [inline plugins/profiles.koplugin/main.lua]: You should also add a `sorting_hint = "tools",`.

PS I haven't tried it yet or thought about where it'd be best to put it.

**Comment by Frenzie** [comment]
> I thought about trying to hook into the gesture manager's actions, as a way to have a selection of relevant actions available instead of arbitrary code execution. However, actions are too limited for my use case, as they don't allow you to select custom values (e.g. there's a "toggle frontlight" action, but there's no "set frontlight intensity to x", nor support in the GUI for choosing x. Similarly, there's a "toggle night mode", but no "turn night mode on", etc.).

Doesn't mean there shouldn't be. ;-)

**Comment by Thra11** [comment]
It should be simple enough to just store a table of parameters (nothing/empty table for actions which don't need a parameter) along with the string which identifies the action. So instead of storing
```"toggle_frontlight"```,
you'd store something like
```{ "set_frontlight_intensity", 50 }```
Lua seems to be quite good at handling things which might not be there: it would be nice if we could leave existing settings untouched.

In addition to the `action_strings`, we'd need a list of parameters that an action takes, with each parameter specifying its type (boolean, integer, enumeration, etc.), valid values (e.g. min & max for integers, possible choices for enumerations). Then the UI could decide what type of control to show, such as a checkbox for a boolean, an input text box for an integer, or a menu for an enumeration.

If we did implement this, then a profile would simply be a list of actions, and loading a profile would consist of iterating over the list triggering each action.

**Comment by poire-z** [comment]
> Then the UI could decide what type of control to show, such as a checkbox for a boolean, an input text box for an integer, or a menu for an enumeration

Just a thought if you feel like going with a UI:
Instead of having a specific UI with again some ways to configure stuff (that we can already configure each in their own place), you could just show a list of checkbox'able settings display name (without their value, and no way to configure them), and the user would just check all the settings he wants included in this profile, and ignore other. You would then just save the ones checked, with their current (not shown to the user) value(s) in your profile file. Like:
```
Settings to include in this new profile
   [x] nighmode
   [x] frontlight level
   [ ] font size
   [x] font contrast
...
```
(No way to edit a current one, except applying it, and just reselect what settings to include, and replace.)

**Comment by Frenzie** [comment]
@Thra11 Btw, other than my minor nits I have no objection to merging this as is for now. :-)

**Comment by Thra11** [comment]
when I have time, I think @poire-z's suggestion of creating profiles by saving the current setup with a filter to select which settings to include is probably the best way to do it, but I haven't had time to implement it so far. In the meantime, I've no objection to it being merged as is (with your changes). It can't hurt anyone by existing, and somebody might find it useful.

I'll leave it to you to decide :-)

**Comment by Frenzie** [comment]
I'm of the opinion that it's easier for others to build upon once it's in. :-)

**Comment by arooni** [comment]
One vote for including this in an upcoming release.

**Comment by Frenzie** [comment]
Effectively superseded by #6106, although some of the functionality implemented here as a proof of concept isn't available in there yet.

**Comment by yparitcher** [comment]
with #6364 plugins should be on par with gestures so this can probably be closed. :)

**Comment by Thra11** [comment]
Haven't had a chance to try it out yet, but sounds good to me!

**Comment by Thra11** [comment]
So having tried out #6106/#6364 in v2020.07.1, it looks there are still a few areas in which it can't quite replicate what this proof of concept did: 
1. Profiles (at least as implemented by this PR) are all about putting certain options in a specific state, which shouldn't depend on the state before the profile was loaded. It is not possible to set the backlight to a specific value using the currently available commands, as there is no "set backlight brightness to x", only "increase frontlight brightness by x" and "decrease frontlight brightness by x". Same problem for backlight warmth.
2. I created a couple of profiles, but I can't seem to trigger profiles via a gesture: Is that supposed to be an option?

**Comment by NiLuJe** [comment]
@Thra11:

1. #6440 ;)
2. @yparitcher mentioned that this was in the pipeline once the groundwork was solid enough ;).

**Comment by yparitcher** [comment]
2. is being implemented in #6441

**Comment by yparitcher** [comment]
Dispatcher is not perfected yet (you can't change every setting in every way yet), but as is seen with #6440 it is very easy just to add a new state to change as long as there is a corresponding handler somewhere in koreader. So if there are any other edge setting missing please open a issue / PR to add it to Dispatcher

**Comment by Thra11** [comment]
> gestures won't call a profile but will natively have all the abilities of a profile.

So is the plan that a gesture can have multiple functions selected?

**Comment by yparitcher** [comment]
> > gestures won't call a profile but will natively have all the abilities of a profile.
> 
> So is the plan that a gesture can have multiple functions selected?

yes

**Comment by Thra11** [comment]
Cool. Awesome work!

**Comment by solarkite** [comment]
I've added a FR issue #8531_
I requested the option to select from available fonts in the profiles menu.
And saving current reading settings under new or existing profiles.

**Pull request closed without merging**